### DISCUSSION TITLE
Using live plugin to write a clion plugin

### DISCUSSION THREAD

**Discussion by mlthlschr** [dkandalov/live-plugin#162] [Q&A]
Hi there,

I have just tried out Live Plugin and it seems to be a great help to start developing plugins. At the moment I am writing code in IntelliJ IDEA, as I have kotlin code completion there. But actually I want to write a plugin for debugging C code. To do this, I need to access the CLion specific APIs, which I cannot access in IDEA. Or do I?

If I switch to CLion, I get even less as the IDEA Kotlin magic is gone and I am left with unhighlighted code and guessing what the API of the CLion stuff might be.

Is there a better way?

**Comment by dkandalov** [comment]
Hi.

Overall, I think you're right.
CLion doesn't have Kotlin support so no highlighting or auto-complete.
There used to be Kotlin/Native plugin for CLion but it's not supported anymore :( https://plugins.jetbrains.com/plugin/10454-kotlin-native-for-clion

Depending on what kind of plugin you want to write you might be able to do it in IntelliJ and then port the code assuming the debugger API is the same/compatible in CLion. If the CLion API you want to use is very different (and I think this means it'll be the proprietary part of CLion), then I'd avoid it altogether because in my experience it's hard to work with non-public API by guessing what it does or decompiling the code. And it will be even harder to support it in the future.

Considering the above, some parts of private API in CLion are probably fine to use, e.g. AST/PSI classes (which you can discover with something like this plugin https://plugins.jetbrains.com/plugin/227-psiviewer, maybe I should embed a similar tool into LivePlugin).

Hope this helps 🙂

Regards,
Dmitry

**Reply by mlthlschr** [reply]
Too bad it is not supported anymore. Looks like another reason for Ultimate (if I get to do some Kotlin native)... 

Anyhow, thanks for the suggestions. I was already wondering why the documentation of the `cidr.*` APIs is so poor.
After reading the docs some more I found out that some of my problems could be solved by modifying `.gdbinit/.lldbinit/NatVis` ([see here](https://www.jetbrains.com/help/clion/configuring-debugger-options.html#gdbinit-lldbinit)). Maybe that's enough for now.

Thanks for the great plugin!
